core/include/detray/propagator/rk_stepper.ipp

```diff
@@ -132,6 +133,49 @@ void detray::rk_stepper<magnetic_field_t, transform3_t, constraint_t, policy_t,
     this->_jac_transport = D * this->_jac_transport;
 }
 
+template <typename magnetic_field_t, typename transform3_t,
+          typename constraint_t, typename policy_t,
+          template <typename, std::size_t> class array_t>
+auto detray::rk_stepper<magnetic_field_t, transform3_t, constraint_t, policy_t,
+                        array_t>::state::evaluate_qop(const scalar h)
+    -> scalar {
+
+    const auto mat = this->_mat;
+    auto& track = this->_track;
+    const scalar qop = track.qop();
+
+    if (mat == detray::vacuum<scalar>()) {
+        return qop;
+    }
+
+    const scalar mass = this->_mass;
+    const auto pdg = this->_pdg;
+    const scalar p = track.p();
+    const scalar q = track.charge();
+    const auto direction = this->_direction;
+
+    scalar new_qop = track.qop();
+
+    if (this->_use_mean_loss) {
+
+        const scalar eloss = interaction<scalar>().compute_energy_loss_bethe(
+            h, mat, pdg, mass, qop, q);
+
+        // @TODO: Recycle the codes in pointwise_material_interactor.hpp
+        // Get new Energy
+        const scalar nextE{
+            std::sqrt(mass * mass + p * p) -
+            std::copysign(eloss, static_cast<scalar>(direction))};
+
+        // Put particle at rest if energy loss is too large
+        const scalar nextP{
+            (mass < nextE) ? std::sqrt(nextE * nextE - mass * mass) : 0.f};
+        new_qop = (q != 0.f) ? q / nextP : 1.f / nextP;
```

Comment:
Won't this cause an FPE when ```mass < nextE``` and hence ```nextP = 0```?

Reply:
You are right. I will just put a assert of (nextE > mass)